File: src/import_data.py
import sys
import argparse
import csv
import datetime
import time
import psycopg2
import string
import os.path

# Retrieving database configuration
config_file = 'database_import.config'
infos = []
with open(config_file) as tsv:
    for line in csv.reader(tsv, delimiter = '\t'):
        infos.append(line[1])
conn_string = "host = '" + infos[0] + "' port = '" + infos[1] + "' dbname = '" + infos[2] + "' user = '" + infos[3] + "' password = '" + infos[4] + "'"

# Opening the connection to the database
dbconn = psycopg2.connect(conn_string)
cursor = dbconn.cursor()

#SQL statements used in the script:
add_assoc = ("INSERT INTO associations "
             "(ID, dataset_id, effect, effect_l95, effect_u95, n, p, p_fdr)"
             "VALUES(%s, %s, %s, %s, %s, %s, %s, %s)")

add_var = ("INSERT INTO variables (ID, label, description) VALUES(%s, %s, %s)")
update_var = ("UPDATE variables SET description = %s WHERE label = %s")
varquery = ("SELECT * FROM variables WHERE label=%s")

add_meta = ("INSERT INTO datasetmetadata (ID, label, description) VALUES (%s, %s, %s)")
metaquery = ("SELECT * FROM datasetmetadata WHERE label = %s")

add_metatodata = ("INSERT INTO datasettometadata (ID, dataset_id, datasetmetadata_id) VALUES (%s, %s, %s)")

add_assoctovar = ("INSERT INTO associationtovariable (ID, association_id, variable_id) VALUES (%s, %s, %s)")

add_dataset = ("INSERT INTO Datasets (ID, label, description, varnum, effect_type) VALUES (%s, %s, %s, %s, %s)")
add_rowcount = ("UPDATE Datasets SET rowcount = %s WHERE id = %s")
datasetquery = ("SELECT * FROM datasets WHERE label=%s")

add_metavar = ("INSERT INTO metavariables (ID, label) VALUES (%s, %s)")
metavarquery = ("SELECT * FROM metavariables WHERE label = %s")
add_numval = ("INSERT INTO numval (ID, value, association_id, metavariable_id) VALUES (%s, %s, %s, %s)")
add_strval = ("INSERT INTO strval (ID, value, association_id, metavariable_id) VALUES (%s, %s, %s, %s)")

# Set all IDs to 1
datasetID = associationID = datasetToMetaDataID = datasetMetaDataID = associationToVariableID = variableID = numvalID = strvalID = metavariableID = 1


def executeSQLFromFile(filename):
    # Open and read the file as a single buffer
    fd = open(filename, 'r')
    sqlFile = fd.read()
    fd.close()

    # all SQL commands (split on ';')
    sqlCommands =  filter(None, sqlFile.strip('\n').split(';'))

    # Execute every command from the input file
    for command in sqlCommands:
        cursor.execute(command)


# Import dataset metadata from file
def import_metadata(filename):
    global datasetMetaDataID
    print( "\n")
    with open(filename,'r') as metafile:
        rdr = csv.DictReader(metafile, delimiter = ",")
        print("Inserting metadata...")
        for row in rdr:
            metadata = (datasetMetaDataID, row["LABEL"],row["DESCRIPTION"])
            cursor.execute(add_meta, metadata)
            datasetMetaDataID += 1

        print("Metadata inserted")


# Get ID of dataset metadata with know label
def get_metaid(meta_label):
    cursor.execute(metaquery, (meta_label,))
    meta_row = cursor.fetchone()
    if(meta_row is not None):
        return meta_row[0]
    else:
        return None


# Populate DatasetToMetaData table
# Link datasets to metadata tags
def import_datatometa(dsid, tagstr):
    global datasetToMetaDataID
    print("Linking to metadata tags...")
    tags = tagstr.split(';')
    count = 0
    for tag in tags:
        metaid = get_metaid(tag)
        if (metaid is not None):
            cursor.execute(add_metatodata, (datasetToMetaDataID, dsid, metaid))
            datasetToMetaDataID += 1
            count += 1
    print(count, " tag(s) associated with the dataset")


# Update description of variables
def update_variables(filename):
    with open(filename,'r') as varfile:
        rdr = csv.DictReader(varfile, delimiter = ",")
        print("Updating variables...")
        count = 0
        for row in rdr:
            cursor.execute(update_var, (row["DESCRIPTION"], row["LABEL"]))
            count += 1

    print("Description of ", count, "variables updated")
   
   
# inserts variable into the table in the database,                  
# sets decription identical to label, returns variable id
def create_varid(var_label):
    global variableID
    # Check if variable with the same label already exists in the database
    cursor.execute(varquery, (var_label,))
    var_row = cursor.fetchone()
    if var_row is not None:
        varID = var_row[0]
    else:
        cursor.execute(add_var, (variableID, var_label,var_label))
        dbconn.commit()
        varID = variableID
        variableID += 1
    return varID


# Populates the AssociationToVariable table
# Links associations to variables
def import_assoctovar(assoc_id, var_label):
    global associationToVariableID
    varid = create_varid(var_label)
    if (varid is not None):
        cursor.execute(add_assoctovar, (associationToVariableID, assoc_id, varid))
        associationToVariableID += 1
        
def import_metavariable(label):
    global metavariableID
    
    if metavar_row is None:
        cursor.execute(add_metavar, (metavariableID, label))
        metavariableID +=1

def isfloat(value):
    try:
        float(value)
        return True
    except ValueError:
        return False
        
def isfloat_vector(v):
    all_float = True
    for value in v:
        if not isfloat(value):
            all_float = False
            break
    return all_float
   
# Read csv file into dict     
def read_to_dict(filename, maxlines):
        header = ''
        data = {}
        with open(filename) as f:
            rdr = csv.reader(f)
            header = next(rdr)
            for i in range(0,len(header)):
                data[header[i]] = []
            rowcount = 0
            for row in rdr:
                for i in range(0,len(row)):
                    data[header[i]].append(row[i])
                    rowcount += 1
                if (maxlines > 0 and rowcount >= maxlines):
                    break
        return data
    
# Adds associations to the database, checks if the dataset has one or two variables and acts accordingly
def import_associations(filename, dsid, varnum, maxlines):
    global metavariableID, associationID, strvalID, numvalID
    
    associations = read_to_dict(filename, maxlines)
    
    ncol = len(associations.keys())
    if varnum == 1:
        normal_columns = ["VARIABLE1", "EFFECT","EFFECT_L95","EFFECT_U95","N","P","P_FDR"]
    else:
        normal_columns = ["VARIABLE1_LABEL","VARIABLE2_LABEL", "EFFECT","EFFECT_L95","EFFECT_U95","N","P","P_FDR"]
    col_limit = len(normal_columns)
    
    # Check if there are extra columns (metavariables)
    # If yes, populate metavariables table and save info about metavariables to a 2D-table
    # First column: metavariable names, second column: type of variable, "num" or "str"
    if ncol > col_limit:
        # Get the names of metavariables
        metavar_names = list(set(associations.keys()) - set(normal_columns))
        metavarnum = ncol - col_limit
        metavars = [None] * metavarnum
        for i in range(0,metavarnum):
            metavars[i] = [None] * 3
            metavar_label = metavar_names[i]
            metavars[i][0] = metavar_label
            # If metavariable already exists in the database, get its ID
            cursor.execute(metavarquery, (metavar_label,))
            metavar_row = cursor.fetchone()
            if metavar_row is not None:
                metavars[i][1] = metavar_row[0]
            # If metavariable does not exist in the database, import it and save ID
            else:
                metavars[i][1] = metavariableID
                cursor.execute(add_metavar, (metavariableID, metavar_label))
                metavariableID +=1
            if isfloat_vector(associations[metavar_label]):
                metavars[i][2] = "num"
            else:
                metavars[i][2] = "str"
        dbconn.commit()
        
    rowcount = 0
    nrow = len(associations["EFFECT"])
    for i in range(0,nrow):
        data_assoc = (associationID, dsid, associations["EFFECT"][i],associations["EFFECT_L95"][i], associations["EFFECT_U95"][i], associations["N"][i], associations["P"][i], associations["P_FDR"][i])
        cursor.execute(add_assoc, data_assoc)
        import_assoctovar(associationID,associations["VARIABLE1_LABEL"][i])
        if varnum == 2:
            import_assoctovar(associationID, associations["VARIABLE2_LABEL"][i])
        if ncol > col_limit:
            for j in range(0,metavarnum):
                if metavars[j][2] == "num":
                    cursor.execute(add_numval,(numvalID, associations[metavars[j][0]][i],associationID,metavars[j][1]))
                    numvalID += 1
                if metavars[j][2] == "str":
                    cursor.execute(add_strval,(strvalID, associations[metavars[j][0]][i],associationID,metavars[j][1]))
                    strvalID += 1
        rowcount += 1
        associationID += 1
        if(maxlines > 0 and rowcount >= maxlines):
            break
        if(rowcount % 1000 == 0):
            dbconn.commit()
            print("{0} rows imported".format(rowcount))

    cursor.execute(add_rowcount, (rowcount, dsid))
    print("{0} rows imported".format(rowcount))


# Read the datasets file, populate the Datasets table and calls other functions to populate
# the other tables
def import_datasets(filename, maxlines):
    global datasetID
    with open(filename,'r') as dataset_file:
        rdr = csv.DictReader(dataset_file, delimiter = ",")
        for row in rdr:
            print("\n")
            dataset = (datasetID, row["LABEL"],row["DESCRIPTION"],row["VARNUM"],row["EFFECT_TYPE"])
            print("Importing dataset ", row["DATASET_FILENAME"], "...")
            cursor.execute(add_dataset, dataset)
            print("Dataset information inserted")
            import_associations(row["DATASET_FILENAME"], datasetID, int(row["VARNUM"]), maxlines)
            import_datatometa(datasetID, row["METADATA_LABELS"])
            if (row["VARIABLES_FILENAME"] is not None and row["VARIABLES_FILENAME"] != ""):
                update_variables(row["VARIABLES_FILENAME"])
            datasetID += 1


class FileNotFoundError(Exception):
    def __init__(self,filename):
        self.file = filename

# Check that all the files containing the associations and variables exist
# If not, throw a FileNotFoundError
def check_files_exist(dsfile):
    with open(dsfile,'r') as dataset_file:
        rdr = csv.DictReader(dataset_file, delimiter = ",")
        for row in rdr:
            df = row["DATASET_FILENAME"]
            if not os.path.isfile(df):
                raise FileNotFoundError(df)
            vf = row["VARIABLES_FILENAME"]
            if (row["VARIABLES_FILENAME"] is not None and row["VARIABLES_FILENAME"] != ""):
                if not os.path.isfile(vf):
                    raise FileNotFoundError(vf)
                    

# Get the next ID number for a table
# next ID = last ID in the table + 1
def get_next_id(table):
    query = "SELECT id FROM " + table + " ORDER BY id DESC LIMIT 1;"
    cursor.execute(query)
    last_row = cursor.fetchone()
    if last_row is not None:
        last_id = last_row[0]
    else:
        last_id = 0
    return last_id + 1

# Set IDs for all tables to the next ID
def set_table_ids():
    global datasetID, associationID, datasetToMetaDataID, datasetMetaDataID, associationToVariableID, variableID, numvalID, strvalID, metavariableID
    datasetID = get_next_id("datasets")
    associationID =  get_next_id("associations")
    datasetToMetaDataID = get_next_id("datasettometadata")
    datasetMetaDataID = get_next_id("datasetmetadata")
    associationToVariableID = get_next_id( "associationtovariable")
    variableID = get_next_id("variables")
    numvalID = get_next_id("numval")
    strvalID = get_next_id("strval")
    metavariableID = get_next_id("metavariables")
        

# Get labels of all datasets to be imported that already exist in the database
def datasets_exist(datasetfile):
    exist = []
    with open(datasetfile,'r') as dataset_file:
        rdr = csv.DictReader(dataset_file, delimiter = ",")
        for row in rdr:
            label = row["LABEL"]
            cursor.execute(datasetquery, (label,))
            dataset_row = cursor.fetchone()
            if dataset_row is not None:
                exist.append(label)
    return exist
            
            
def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("-dsf", "--dataset_file", help="The file containing information on the datasets to import")
    parser.add_argument("-mdf","--meta_data_file", help="The file containing the metadata for the datasets")
    parser.add_argument("-a","--append", action = 'store_true', help="Add dataset to the databse instead of clearing the database before import")
    parser.add_argument("-ml","--maxlines", default = -1, help = "Maximum number of lines imported per dataset (mainly for testing, defaults to unlimited)")
    
    args = parser.parse_args()
    
    if not args.append: # Database is cleared before import
        # Drop and recreate the database schema
        executeSQLFromFile("drop_schema.sql")
        print("Database schema dropped")
        executeSQLFromFile("create_schema.sql")
        print("Database schema recreated")
        
    if args.append: # New datasets will be appended to the database
        # If any of the datasets to be appended already exist in the database, stop the script
        existing_datasets = datasets_exist(args.dataset_file)
        if len(existing_datasets) > 0:
            print("Error: Following datasets already exist in the database:")
            for e in existing_datasets:
                print(e)
            print("Please only apend datasets that are not found in the database or remove --append to clear database before import")
            dbconn.commit()
            cursor.close()
            dbconn.close()
            sys.exit()
        # Set IDs for all the tables to next logical ID number
        set_table_ids()
        
    # Import dataset metadata
    if args.meta_data_file:
        import_metadata(args.meta_data_file)

    # Import datasets
    if args.dataset_file:
        try:
            check_files_exist(args.dataset_file)
            import_datasets(args.dataset_file, int(args.maxlines))
        except FileNotFoundError as e:
            print("File not found:", e.file)
    
    # Close connection to the database                     
    dbconn.commit()
    cursor.close()
    dbconn.close()

if __name__ == "__main__":
    main()

File: src/import_data.R
library(DBI)

setwd("~/UEF_projects/ninni/")
source("app/www/db_functions.R")

db_info <- read_db_info("app/www/database_www.config")

con <- dbConnect(
  drv = RPostgres::Postgres(),
  dbname = db_info$db_name,
  host = db_info$db_host,
  port = db_info$db_port,
  user = db_info$db_user,
  password = db_info$db_password)

# orig <- list()
# for (tbl in dbListTables(con)) {
#   orig[[tbl]] <- dbReadTable(con, tbl)
# }
# 
# save(orig, file = "app/orig_data.RData")



# Real code
append <- FALSE


if (append) {
  
  # Check if datasets exist in database, skip those
  
  ids <- get_last_ids(con)
} else {
  execute_sql_file("src/drop_schema.sql", con)
  cat("Dropped database schema\n")
  execute_sql_file("src/create_schema.sql", con)
  cat("Created database schema\n")
  
  # Initialize all last IDs to 0
  ids <- sapply(dbListTables(con), function(x) {0})
}


# Write old data back for demoing Ninni
# load(file = "app/orig_data.RData")
# for (tbl in names(orig)) {
#   tryCatch({
#     dbAppendTable(con, tbl, orig[[tbl]])
#   }, error = function(e) {print(e$message)})
#   
# }

# Helper function for appending data to database
append_table <- function(df, tbl) {
  if (!"id" %in% colnames(df)) {
    df$id <- seq_len(nrow(df)) + ids[tbl]
  }
  colnames(df) <- tolower(colnames(df))
  dbAppendTable(con, tbl, df)
}

# Import dataset metadata
metadata_old <- dbReadTable(con, "datasetmetadata")
metadata <- read.csv("example_data/metadata.csv", stringsAsFactors = FALSE)
colnames(metadata) <- tolower(colnames(metadata))

# Remove metadata labels already in the database
metadata <- metadata[!metadata$label %in% metadata_old$label, ]
append_table(metadata, "datasetmetadata")

# Import datasets
datasets <- read.csv("example_data/datasets.csv", stringsAsFactors = FALSE)
datasets$DATASET_FILENAME <- gsub("[.][.]/", "", datasets$DATASET_FILENAME)
datasets$VARIABLES_FILENAME <- gsub("[.][.]/", "", datasets$VARIABLES_FILENAME)

# Check that files exist
check_files(datasets)

# Import datasets
colnames(datasets) <- tolower(colnames(datasets))

assocs_full <- lapply(datasets$dataset_filename, read.csv, stringsAsFactors = FALSE) 

datasets$rowcount <- sapply(assocs_full, nrow)
datasets$id <- seq_len(nrow(datasets)) + ids["datasets"]
dbAppendTable(con, "datasets", datasets[c("id", "label", "rowcount", "description", "varnum", "effect_type")])

# Dataset to metadata
# Read the whole metadata table
metadata <- dbReadTable(con, "datasetmetadata")
ds2md <- data.frame()
for (i in seq_len(nrow(datasets))) {
  meta_labels <- strsplit(datasets$metadata_labels[i], split = ";")[[1]]
  for (label in meta_labels) {
    ds2md <- rbind(ds2md, data.frame(datasetmetadata_id = metadata$id[metadata$label == label],
                                     dataset_id = datasets$id[i]))
  }
}
append_table(ds2md, "datasettometadata")


# Import associations
variables_old <- variables_all <-  dbReadTable(con, "variables")
metavariables_old <- metavariables_all <- dbReadTable(con, "metavariables")
# Initialize new data frames to import
variables <- metavariables <- associations <- assoc2var <- numval <- strval <- data.frame()

# Add stuff to be imported from each dataset
for (i in seq_len(nrow(datasets))) {
  # Read variable descriptions
  var_fname <- datasets$variables_filename[i]
  if (i == 2) var_fname <- ""
  if (!is.na(var_fname) && var_fname != "") {
    variables_tmp <- read.csv(var_fname, stringsAsFactors = FALSE)
    colnames(variables_tmp) <- tolower(colnames(variables_tmp))
  } else {
    variables_tmp <- data.frame()
  }
  
  # Find variables with only a dummy description in the database
  dummy_vars <- variables_old[variables_old$label %in% variables$label, ]
  dummy_vars <- dummy_vars[dummy_vars$label == dummy_vars$description, ]
  if (nrow(dummy_vars)) {
    cat(paste("Updating the description of", nrow(dummy_vars), "variables\n"))
    dummy_vars$description <- NULL
    dummy_vars <- dplyr::left_join(dummy_vars, variables, by = "label")
    # Update their descriptions with the new descriptions
    for (j in seq_len(nrow(dummy_vars))) {
      dbSendQuery(con, statement = paste0("UPDATE variables set description = '", dummy_vars$description[j],
                                          "' WHERE id = ", dummy_vars$id[j]))
    }
  }
  # Define usual columns and list of current variables
  assoc <- assocs_full[[i]]
  varnum <- datasets$varnum[i]
  if (varnum == 1){
    normal_columns = c("VARIABLE1_LABEL", "EFFECT", "EFFECT_L95", "EFFECT_U95", "N", "P", "P_FDR")
    # Variables to import
    var_labels <- unique(assoc$VARIABLE1_LABEL)
  } else {
    normal_columns = c("VARIABLE1_LABEL", "VARIABLE2_LABEL", "EFFECT", "EFFECT_L95", "EFFECT_U95", "N", "P", "P_FDR")
    # Variables to import
    var_labels <- unique(c(assoc$VARIABLE1_LABEL, assoc$VARIABLE2_LABEL))
  }
  
  # Add dummy descriptions for variables without actual description
  new_vars <- setdiff(var_labels, variables_tmp$label)
  variables_tmp <- rbind(variables_tmp, data.frame(label = new_vars, description = new_vars,
                                           stringsAsFactors = FALSE))
  # Remove variables already in the database and add them to the variables to be appended
  variables_tmp <- variables_tmp[!variables_tmp$label %in% variables_all$label, ]
  variables_tmp$id <- seq_len(nrow(variables_tmp)) + ids["variables"]
  ids["variables"] <- ids["variables"] + nrow(variables_tmp)
  variables <- rbind(variables, variables_tmp)
  variables_all <- rbind(variables_all, variables_tmp)
  # Take IDs for all variables in current set, even previously found
  # These are needed when linking variables with associations
  variables_tmp <- variables_all[variables_all$label %in% var_labels, ]
  rownames(variables_tmp) <- variables_tmp$label
  
  # New associations
  
  associations_tmp <- assoc[normal_columns]
  associations_tmp$Dataset_ID <- datasets$id[i]
  associations_tmp$id <- seq_len(nrow(associations_tmp)) + ids["associations"]
  ids["associations"] <- ids["associations"] + nrow(associations_tmp)
  rownames(assoc) <- associations_tmp$id
  associations <- rbind(associations, associations_tmp[c("id", "Dataset_ID", "EFFECT", "EFFECT_L95", "EFFECT_U95", "N", "P", "P_FDR")])
  # Link associations to variables
  assoc2var_tmp <- data.frame(association_id = associations_tmp$id,
                              variable_id = variables_tmp[associations_tmp$VARIABLE1_LABEL, "id"])
  if (varnum == 2) {
    assoc2var_tmp <- rbind(assoc2var_tmp,
                           data.frame(association_id = associations_tmp$id,
                                      variable_id = variables_tmp[associations_tmp$VARIABLE2_LABEL, "id"]))
  }
  assoc2var <- rbind(assoc2var, assoc2var_tmp)
  
  # Check if there are extra columns (metavariables)
  # If yes, populate metavariables table and save types metavariables to a named vector
  if (ncol(assoc) > length(normal_columns)) {
    # Get the names of metavariables
    metavar_labels <- setdiff(colnames(assoc), normal_columns)
    # Add new metavariables to metavariables table
    metavariables_tmp <- data.frame(label = setdiff(metavar_labels, metavariables_all$label),
                                    stringsAsFactors = FALSE)
    metavariables_tmp$id <- seq_len(nrow(metavariables_tmp)) + ids["metavariables"]
    ids["metavariables"] <- ids["metavariables"] + nrow(metavariables_tmp)
    metavariables <- rbind(metavariables, metavariables_tmp)
    metavariables_all <- rbind(metavariables_all, metavariables_tmp)
    # Take IDs for all metavariables in current set, even previously found
    metavariables_tmp <- metavariables_all[metavariables_all$label %in% metavar_labels, ]
    rownames(metavariables_tmp) <- metavariables_tmp$label
    # Determine types of metavariables in this dataset
    # NOTE: same metavariable can have different types of values in different datasets
    metavariables_tmp$type <- sapply(metavar_labels, function(x) {
      if (is.numeric(assoc[, x])) {
        "num"
      } else {
        "str"
      }
    })
    # Link associations to metavariables
    numval <- rbind(numval, link_metavars(associations_tmp, assoc, type = "num"))
    strval <- rbind(strval, link_metavars(associations_tmp, assoc, type = "str"))
  }
  
}

# Append variables
dfs <- list(variables = variables,
            metavariables = metavariables,
            associations = associations,
            associationtovariable = assoc2var,
            numval = numval,
            strval = strval)
for (i in seq_along(dfs)) {
  append_table(dfs[[i]], names(dfs)[i])
}


dbDisconnect(con)
